### ISSUE TITLE
Differnt behaviors when setting group_by_lenth to differnt value.

### ISSUE THREAD

**Issue by s1530129650** [tloen/alpaca-lora#335]
I run the program on 4*V100.

When setting group_by_lenth=True, the training loss often decreases and coverages. 

![image](https://user-images.githubusercontent.com/34572118/231823032-6e30ad44-b1c8-4fbe-8531-3784f5dc944d.png)

When setting  group_by_lenth=False, the training loss often becomes suddenly bigger and then it goes to 0. 

![image](https://user-images.githubusercontent.com/34572118/231822678-6f8aed48-68a0-4502-8153-e3d02c0518f4.png)

I know the ` group_by_lenth` would affect the performance. However, it is so wired that the training loss would suddenly get  bigger and then it go to 0 when ` group_by_lenth=True`.

FYI

Running scripts

```
WORLD_SIZE=${number_card}  torchrun --nproc_per_node=${number_card} --master_port=13214  finetune.py \
    --base_model 'decapoda-research/llama-7b-hf'  \
    --data_path alpaca_data.json \
    --output_dir ${output_dir} \
    --batch_size 128 \
    --micro_batch_size 4 \
    --num_epochs  10 \
    --learning_rate 3e-4 \
    --cutoff_len 512 \
    --val_set_size 0 \
    --lora_r 8 \
    --lora_alpha 16 \
    --lora_dropout 0.05 \
    --lora_target_modules '[q_proj,v_proj]'  \
    --train_on_inputs \
    --group_by_length \
```

Env
```
Ubuntu 18
DGX 4*V100-32G
```
